src/abstractdataframe/abstractdataframe.jl

```diff
@@ -969,45 +969,46 @@ julia> filter(AsTable(:) => nt -> nt.x == 1 || nt.y == "b", df)
 │ 3   │ 1     │ b      │
 ```
 """
-@inline Base.filter(f, df::AbstractDataFrame; view::Bool=false) =
-    _filter_helper(df, f, eachrow(df), view=view)
-@inline Base.filter((col, f)::Pair{<:ColumnIndex}, df::AbstractDataFrame;
-    view::Bool=false) = _filter_helper(df, f, df[!, col], view=view)
-@inline Base.filter((cols, f)::Pair{<:AbstractVector{Symbol}},
-                    df::AbstractDataFrame; view::Bool=false) =
-    filter([index(df)[col] for col in cols] => f, df, view=view)
-@inline Base.filter((cols, f)::Pair{<:AbstractVector{<:AbstractString}},
-                    df::AbstractDataFrame; view::Bool=false) =
-    filter([index(df)[col] for col in cols] => f, df, view=view)
-@inline Base.filter((cols, f)::Pair, df::AbstractDataFrame; view::Bool=false) =
-    filter(index(df)[cols] => f, df, view=view)
-@inline Base.filter((cols, f)::Pair{<:AbstractVector{Int}},
-                    df::AbstractDataFrame; view::Bool=false) =
-    _filter_helper(df, f, (df[!, i] for i in cols)...; view=view)
-
-@inline function _filter_helper(df::AbstractDataFrame, f, cols...; view::Bool)
+@inline function Base.filter(f, df::AbstractDataFrame; view::Bool=false)
+    rowidxs::BitVector = _filter_helper(f, eachrow(df))
```

Comment:
It's weird that inference fails here. Anyway, maybe better put the type assertion directly in `_filter_helper` itself?

Reply:
Actually I have tried this and this was crashing my Julia earlier. But now it seems to work, so probably I had some weird problem earlier. I pushed the fix for this. Also I think it is not a problem to add this conversion and assertion as it is no-op and at least we signal what we intend to return.

Reply:
OK. Do we have tests to catch an inference failure if somebody removed the assertion?

Reply:
Now I remember why I did it the other way - it was because nightly fails this way, see e.g. https://travis-ci.org/github/JuliaData/DataFrames.jl/jobs/725236829 (I have not tested this in 100% detail to nail down the reason as I have a lot of classes this week :weary:).

@JeffBezanson + @Keno : this is a regression on nightly vs Julia 1.5.1 - some problems with type inference. Please let me know if you need more information to investigate.

Use this branch on Julia nightly and run the following to reproduce (interestingly - a correct result is produced, but an error is thrown in the process):
```
julia> using DataFrames

julia> df = DataFrame(rand(10,1))
10×1 DataFrame
│ Row │ x1        │
│     │ Float64   │
├─────┼───────────┤
│ 1   │ 0.389567  │
│ 2   │ 0.0901768 │
│ 3   │ 0.179161  │
│ 4   │ 0.0274051 │
│ 5   │ 0.696103  │
│ 6   │ 0.321395  │
│ 7   │ 0.476313  │
│ 8   │ 0.628184  │
│ 9   │ 0.610729  │
│ 10  │ 0.048924  │

julia> filter(:x1 => >(0.5), df)
Internal error: encountered unexpected error in runtime:
BoundsError(a=Array{UInt64, (68,)}[0x0000000000000001, 0x0000000000000002, 0x0000000000000000, 0x0000000000000003, 0x0000000000000004, 0x0000000000000005, 0x0000000000000006, 0x0000000000000007, 0x0000000000000008, 0x0000000000000009, 0x000000000000000a, 0x000000000000000b, 0x0000000000000000, 0x0000000000000000, 0x0000000000000000, 0x0000000000000000, 0x0000000000000000, 0x0000000000000000, 0x0000000000000000, 0x0000000000000000, 0x0000000000000000, 0x0000000000000000, 0x0000000000000000, 0x0000000000000000, 0x0000000000000000, 0x0000000000000000, 0x0000000000000000, 0x0000000000000000, 0x0000000000000000, 0x0000000000000000, 0x0000000000000000, 0x0000000000000000, 0x0000000000000000, 0x0000000000000000, 0x0000000000000000, 0x0000000000000000, 0x0000000000000000, 0x0000000000000000, 0x000000000000000c, 0x000000000000000d, 0x000000000000000e, 0x0000000000000000, 0x0000000000000000, 0x0000000000000000, 0x000000000000000f, 0x0000000000000010, 0x0000000000000011, 0x0000000000000012, 0x0000000000000013, 0x0000000000000000, 0x0000000000000014, 0x0000000000000015, 0x0000000000000016, 0x0000000000000017, 0x0000000000000000, 0x0000000000000018, 0x0000000000000019, 0x000000000000001a, 0x000000000000001b, 0x0000000000000000, 0x000000000000001c, 0x000000000000001d, 0x000000000000001e, 0x000000000000001f, 0x0000000000000020, 0x0000000000000021, 0x0000000000000022, 0x0000000000000023], i=(69,))
jl_bounds_error_ints at /buildworker/worker/package_linux64/build/src/rtutils.c:183
getindex at ./array.jl:809 [inlined]
getindex at ./abstractarray.jl:1122 [inlined]
DFS at ./compiler/ssair/domtree.jl:197
SNCA at ./compiler/ssair/domtree.jl:269
construct_domtree at ./compiler/ssair/domtree.jl:121
run_passes at ./compiler/ssair/driver.jl:131
optimize at ./compiler/optimize.jl:172
typeinf at ./compiler/typeinfer.jl:32
typeinf_edge at ./compiler/typeinfer.jl:522
abstract_call_method at ./compiler/abstractinterpretation.jl:453
abstract_call_gf_by_type at ./compiler/abstractinterpretation.jl:129
abstract_call_known at ./compiler/abstractinterpretation.jl:991
abstract_call at ./compiler/abstractinterpretation.jl:1014
abstract_call at ./compiler/abstractinterpretation.jl:998
abstract_eval_statement at ./compiler/abstractinterpretation.jl:1119
typeinf_local at ./compiler/abstractinterpretation.jl:1375
typeinf_nocycle at ./compiler/abstractinterpretation.jl:1431
typeinf at ./compiler/typeinfer.jl:12
typeinf_ext at ./compiler/typeinfer.jl:609
typeinf_ext_toplevel at ./compiler/typeinfer.jl:642
typeinf_ext_toplevel at ./compiler/typeinfer.jl:638
jfptr_typeinf_ext_toplevel_10082.clone_1 at /home/bkamins/Downloads/julia-latest-linux64/julia-371bfa89d4/lib/julia/sys.so (unknown line)
jl_apply at /buildworker/worker/package_linux64/build/src/julia.h:1750 [inlined]
jl_type_infer at /buildworker/worker/package_linux64/build/src/gf.c:300
jl_generate_fptr at /buildworker/worker/package_linux64/build/src/jitlayers.cpp:293
jl_compile_method_internal at /buildworker/worker/package_linux64/build/src/gf.c:1890
jl_compile_method_internal at /buildworker/worker/package_linux64/build/src/gf.c:1841 [inlined]
_jl_invoke at /buildworker/worker/package_linux64/build/src/gf.c:2145 [inlined]
jl_apply_generic at /buildworker/worker/package_linux64/build/src/gf.c:2336
jl_apply at /buildworker/worker/package_linux64/build/src/julia.h:1750 [inlined]
do_call at /buildworker/worker/package_linux64/build/src/interpreter.c:117
eval_value at /buildworker/worker/package_linux64/build/src/interpreter.c:206
eval_stmt_value at /buildworker/worker/package_linux64/build/src/interpreter.c:157 [inlined]
eval_body at /buildworker/worker/package_linux64/build/src/interpreter.c:551
jl_interpret_toplevel_thunk at /buildworker/worker/package_linux64/build/src/interpreter.c:659
top-level scope at REPL[15]:1
jl_toplevel_eval_flex at /buildworker/worker/package_linux64/build/src/toplevel.c:838
jl_toplevel_eval_flex at /buildworker/worker/package_linux64/build/src/toplevel.c:788
jl_toplevel_eval_in at /buildworker/worker/package_linux64/build/src/toplevel.c:881
eval at ./boot.jl:344
eval_user_input at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.6/REPL/src/REPL.jl:134
repl_backend_loop at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.6/REPL/src/REPL.jl:195
start_repl_backend at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.6/REPL/src/REPL.jl:180
#run_repl#41 at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.6/REPL/src/REPL.jl:311
run_repl at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.6/REPL/src/REPL.jl:299
#844 at ./client.jl:386
jfptr_YY.844_30029.clone_1 at /home/bkamins/Downloads/julia-latest-linux64/julia-371bfa89d4/lib/julia/sys.so (unknown line)
jl_apply at /buildworker/worker/package_linux64/build/src/julia.h:1750 [inlined]
do_apply at /buildworker/worker/package_linux64/build/src/builtins.c:655
jl_f__apply_latest at /buildworker/worker/package_linux64/build/src/builtins.c:705
#invokelatest#2 at ./essentials.jl:718 [inlined]
invokelatest at ./essentials.jl:717 [inlined]
run_main_repl at ./client.jl:371
exec_options at ./client.jl:301
_start at ./client.jl:484
jfptr__start_22499.clone_1 at /home/bkamins/Downloads/julia-latest-linux64/julia-371bfa89d4/lib/julia/sys.so (unknown line)
jl_apply at /buildworker/worker/package_linux64/build/ui/../src/julia.h:1750 [inlined]
true_main at /buildworker/worker/package_linux64/build/ui/repl.c:106
main at /buildworker/worker/package_linux64/build/ui/repl.c:227
__libc_start_main at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
_start at /home/bkamins/Downloads/julia-latest-linux64/julia-371bfa89d4/bin/julia (unknown line)
3×1 DataFrame
│ Row │ x1       │
│     │ Float64  │
├─────┼──────────┤
│ 1   │ 0.696103 │
│ 2   │ 0.628184 │
│ 3   │ 0.610729 │
```

Reply:
> Do we have tests to catch an inference failure if somebody removed the assertion?

Yes - tests will catch this then.